fence/man/fence_scsi.8
.TH fence_scsi 8

.SH NAME
fence_scsi - I/O fencing agent for SCSI persistent reservations

.SH SYNOPSIS
.B 
fence_scsi
[\fIOPTION\fR]...

.SH DESCRIPTION
fence_scsi is an I/O fencing agent which can be used with the SCSI
devices that support persistent reservations (SPC-2 or greater).

SCSI persistent reservations work by having each node in the cluster
register with the SCSI device. Registration is done using a unique key
(based on the node's IP address). Each node that will perform I/O
operations to the shared storage must register with the device. This
is done at system startup with the scsi_reserve init script. This
script will discover all clustered volumes as well as the underlying
SCSI device(s). Device discovery is done via the lvs command (see
lvs(8)) and is subject to any filtering rules defined in the lvm.conf
file.

After generating the node's unique key, the script will register the
node with the SCSI device(s) that were discovered. Once the node is
registered, the script will attempt to create a reservation. Unlike
registrations, of which there are multiple registrants (one for each
node in the cluster), there is only one reservation holder. If a
reservation does not already exist for a device, the script will
create a reservation using the node's unique key.

It is important to distinguish between registrations and
reservations. As mentioned above, each node that will perform I/O
operations to the SCSI device must register. As a result, there will
be multiple registrations for a given SCSI device. In contrast, there
can only be one reservation per SCSI device. It is not important which
node holds the reservation. The reservation simply tells the device
how the registrants are allowed to access the device. For our
purposes, the reservation type is "write exclusive, registrants only".
With this reservation type, only registered nodes will be able to
write to the device.

When the cluster must fence a node, it simply revokes a node's
registration, or "unregisters" the node. This operation also uses the
node's unique key. By deriving the unique key based on the errant
node's IP address, the cluster can "unregister" the key. As a
result, the errant node will no longer be able to write to the SCSI
device.

Note that the node that holds the reservation for a device must also
be registered with that device. When the situation arises where the
node that is being fenced is also the reservation holder, the
reservation must be moved. This is handled by using the
"preempt-and-abort command" which will transfer the reservation from
the node that is being fenced to the node that is performing the
fencing. This operation will maintain the reservation while
"unregistering" the node being fenced.

At system shutdown, the scsi_reserve script will attempt to
"unregister" the node from all devices. The exception is when the
node happens to be the reservation holder. In this case, the script
does nothing, due to the fact that there may be other nodes using the
device and the reservation must remain intact.

fence_scsi accepts options on the command line as well as from stdin.
fenced sends parameters through stdin when it execs the agent.  fence_scsi
can be run by itself with command line options.  This is useful for testing 
and for turning outlets on or off from scripts.

.SH OPTIONS
.TP
\fB-n\fP \fInode\fR
Name of the node to be fenced.
.TP
\fB-h\fP
Print out a help message describing available options, then exit.
.TP
\fB-s\fP \fIself\fR
Name of the node that will perform the fencing operation.
.TP
\fB-v\fP
Verbose output.
.TP
\fB-V\fP
Print out a version message, then exit.

.SH STDIN PARAMETERS
.TP
\fIagent = < param >\fR
This option is used by fence_node(8) and is ignored by fence_scsi.
.TP
\fInodename = < hostname | ip >\fR
Name of the node to be fenced.
.TP
\fIself = < nodename >\fR
Name of the node that will perform the fencing operation.
.TP
\fIverbose = < param >\fR
Verbose output.

.SH SEE ALSO
fence(8), fence_node(8), sg_persist(8), lvs(8), lvm.conf(5)


fence/agents/scsi/fence_scsi.pl
#!/usr/bin/perl

use Getopt::Std;
use IPC::Open3;
use POSIX;

my $verbose = 0;
my @volumes;

$_ = $0;
s/.*\///;
my $pname = $_;

# WARNING!! Do not add code bewteen "#BEGIN_VERSION_GENERATION" and
# "#END_VERSION_GENERATION"  It is generated by the Makefile
                                                                                
#BEGIN_VERSION_GENERATION
$RELEASE_VERSION="";
$REDHAT_COPYRIGHT="";
$BUILD_DATE="";
#END_VERSION_GENERATION

sub usage
{
    print "Usage\n";
    print "\n";
    print "$pname [options]\n";
    print "\n";
    print "Options\n";
    print "  -n <node>        IP address or hostname of node to fence\n";
    print "  -h               usage\n";
    print "  -V               version\n";
    print "  -v               verbose\n";

    exit 0;
}

sub version
{
    print "$pname $RELEASE_VERSION $BUILD_DATE\n";
    print "$REDHAT_COPYRIGHT\n" if ( $REDHAT_COPYRIGHT );

    exit 0;
}

sub fail
{
    ($msg)=@_;

    print $msg."\n" unless defined $opt_q;

    exit 1;
}

sub fail_usage
{
    ($msg)=@_;

    print STDERR $msg."\n" if $msg;
    print STDERR "Please use '-h' for usage.\n";

    exit 1;
}

sub get_key
{
    ($node)=@_;

    my $addr = gethostbyname($node) or die "$!\n";

    return unpack("H*", $addr);
}

sub get_options_stdin
{
    my $opt;
    my $line = 0;

    while (defined($in = <>))
    {
	$_ = $in;
	chomp;

	# strip leading and trailing whitespace
	s/^\s*//;
	s/\s*$//;

	# skip comments
	next if /^#/;

	$line += 1;
	$opt = $_;

	next unless $opt;

	($name, $val) = split /\s*=\s*/, $opt;

	if ($name eq "")
	{
	    print STDERR "parse error: illegal name in option $line\n";
	    exit 2;
	}
	elsif ($name eq "agent")
	{
	}
	elsif ($name eq "node")
	{
	    $opt_n = $val;
	}
	elsif ($name eq "self")
	{
	    $opt_s = $val;
	}
	elsif ($name eq "verbose")
	{
	    $opt_v = $val;
	}
	else
	{
	    fail "parse error: unknown option \"$opt\"";
	}
    }
}

sub get_scsi_devices
{
    my ($in, $out, $err);
    my $cmd = "lvs --noheadings --separator : -o vg_attr,devices";
    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "Unable to execute lvs.\n" if ($?>>8);

    while (<$out>)
    {
	chomp;
	print "OUT: $_\n" if $opt_v;

	my ($vg_attrs, $device) = split /:/, $_, 3;

	if ($vg_attrs =~ /.*c$/)
	{
	    $device =~ s/\(.*\)//;
	    push @volumes, $device;
	}
    }

    close($in);
    close($out);
    close($err);
}

sub check_sg_persist
{
    my ($in, $out, $err);
    my $cmd = "sg_persist -V";
    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "Unable to execute sg_persist.\n" if ($?>>8);

    while (<$out>)
    {
	chomp;
	print "OUT: $_\n" if $opt_v;
    }

    close($in);
    close($out);
    close($err);
}

sub fence_node
{
    # my $name = (POSIX::uname())[1];

    my $host_key = get_key($opt_s);
    my $node_key = get_key($opt_n);
    
    my $cmd;
    my ($in, $out, $err);

    foreach $dev (@volumes)
    {
	if ($host_key eq $node_key)
	{
	    $cmd = "sg_persist -d $dev -o -G -K $host_key -S 0";
	}
	else
	{
	    $cmd = "sg_persist -d $dev -o -A -K $host_key -S $node_key -T 5";
	}

	my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

	waitpid($pid, 0);

	if ($opt_v)
	{
	    print "$cmd\n";
	    while (<$out>)
	    {
		chomp;
		print "OUT: $_\n";
	    }
	}

	die "Unable to execute sg_persist.\n" if ($?>>8);

	close($in);
	close($out);
	close($err);
    }
}

### MAIN #######################################################

if (@ARGV > 0) {

    getopts("n:hqvV") || fail_usage;

    usage if defined $opt_h;
    version if defined $opt_V;

    fail_usage "Unkown parameter." if (@ARGV > 0);
    fail_usage "No '-n' flag specified." unless defined $opt_n;
    fail_usage "No '-s' flag specified." unless defined $opt_s;

} else {

    get_options_stdin();

    fail "failed: missing 'node'" unless defined $opt_n;
    fail "failed: missing 'self'" unless defined $opt_s;

}

check_sg_persist;

get_scsi_devices;

fence_node;
